[Requirement Machine] Implement same-element requirements.

simanerush opened the pull request:
Resolves rdar://131630764.

This is the continuation of the original PR #67465 since I hope to fix the CI errors to be able to land the changes.

- [x] Fix CI errors
- [x] Address review feedback
- [ ] Support concrete same-type requirements?

From the original PR:
SE-0393 supports _same-element requirements_. From the proposal:

> A same-type requirement where one side is a type parameter pack and the other type is a scalar type that does not capture any type parameter packs is interpreted as constraining each element of the replacement type pack to the same scalar type:
> ```swift
> func variadic<each S: Sequence, T>(_: repeat each S) where repeat (each S).Element == T {}
> ```
> This is called a _same-element requirement._

We represent pack elements in the requirement machine through a new singleton `PackElement` symbol, written `[element]`. Same-element requirements are mapped to re-write rules that prefix type parameter pack terms with the `[element]` symbol.

For example, given a protocol `P` with an associated type `A`, the generic signature `<each T, U> where repeat each T: P, repeat (each T).A == U` maps to the following re-write system:

```
Rewrite system: {
- [P].[P] => [P]
- [P].A => [P:A]
- [P].[P:A] => [P:A]
- [P:A].[shape] => [shape]
- τ_0_1.[shape] => [shape]
- each τ_0_0.[P] => each τ_0_0
- [element].each τ_0_0.A => τ_0_1
- each τ_0_0.A => each τ_0_0.[P:A]
- [element].each τ_0_0.[P:A] => τ_0_1
- [element].each τ_0_0.[shape] => [shape]
}
```

The implementation is gated behind `-enable-experimental-feature SameElementRequirements` because the minimal conformance computation doesn't yet support same-element requirements. The current rewrite rules also do not support deriving pack element conformance from a scalar type parameter conformance through same-element requirements. For example, the generic signature `<each T, U> where U: P, repeat each T == U` maps to the following rewrite system:

```
Rewrite system: {
- [P].[P] => [P]
- [P].A => [P:A]
- [P].[P:A] => [P:A]
- [P:A].[shape] => [shape]
- τ_0_1.[shape] => [shape]
- τ_0_1.[P] => τ_0_1
- [element].each τ_0_0 => τ_0_1
- τ_0_1.A => τ_0_1.[P:A]
}
```

The conformance rule `τ_0_1.[P] => τ_0_1` and the same-element rule `[element].each τ_0_0 => τ_0_1 ` do not provide a rewrite path to the pack conformance rule `each τ_0_0.[P] => each τ_0_0`. One possible solution is to use element symbols for all property rules, which would also allow the minimal conformance support to fallout naturally. However, this approach poses a problem for same-type-pack requirements. If we use `[element]` symbols for same-type-pack requirements, e.g. `[element].each τ_0_0 => [element].each τ_0_1`, then this rule no longer implies a same-shape rule `each τ_0_0.[shape] => each τ_0_1.[shape]`. Stripping the `[element]` symbols from same-type-pack rewrite rules creates a different set of problems for conformances, because the pack conformance rules contain `[element]` symbols, which would prohibit applying associated type rewrite rules to non-element pack terms. I'm still exploring strategies for resolving these issues, hence the experimental flag!

hborla committed:
[Requirement Machine] Implement same-element requirements.

hborla committed:
[Requirement Machine] Fix the order of requirements that involve pack elements.

hborla committed:
[Requirement Machine] Use element symbols for concrete same-element requirements.

hborla committed:
[Constraint System] Fix the shape class and context substitiutions for
opened element generic environments containing same-element requirements.

hborla committed:
[Requirement Machine] Stricter verification conditions for element symbols.

hborla committed:
[Features] Gate same-element requirements behind an experimental feature flag.

hborla committed:
[NFC] Add -dump-input=fail to a FileCheck test.

simanerush commented:
@swift-ci Please smoke test macOS

slavapestov commented on lib/AST/GenericSignature.cpp:
```diff
@@ -938,7 +946,12 @@ void GenericSignature::verify(ArrayRef<Requirement> reqts) const {
+        // which turns into a same-type requirement repeat U == each T. The
+        // pack element must always be on the left side of the rule, so type
+        // parameters can appear out of order in the final same-type requirement.
+        if (firstType->isParameterPack() == secondType->isParameterPack() &&
```
Don't forget to remove this part -- we can't change type parameter order because it will break ABI

hborla replied:
Sima and I were discussing this and we're a bit confused. This code is just in `GenericSignature::verify`; I don't think this changes type parameter order of existing code. My original implementation did change the order initially but I had fixed this in https://github.com/apple/swift/pull/67465/commits/d138b1553f776f4ae79cea8484c3129fc7d3c4a9.

For a same-element requirement, my understanding is we need the element parameter to be on the left because the reduced type of a pack with a same-element requirement has to be the scalar type, which is why we have this special handling for the order. I see in my original PR you left a comment that the requirement builder should reorder them. Is your suggestion that the rewrite system should order the element terms before pack terms, but we should change that when covering it to a generic signature?

hborla replied:
Okay Slava helped me remember what's going on here: We can fix the `getReducedType` query to always return the scalar type as the reduced type with a dedicated `getReducedElementType` query that's only ever called on the pattern of a pack expansion. We don't necessarily want `getReducedType` to have this behavior for packs in general, because shape types are also represented with parameter packs, which should not be reduced to a scalar type. So we don't necessarily need the minimized generic signature to have this order; we just need the requirement machine and `getReducedElementType` query to handle that. The reduced type query already handles pack expansion types specially to call the `getReducedShape` query, so that's where we'd need to call `getReducedElementType` too.

slavapestov replied:
It would be simplest to maintain the invariant that for a same-type requirement `T == U` we have T < U, so that means a same-element requirement is either going to have the pack on the left or the right depending on the order of the two type parameters. In the reduction order though, the term `[element].T` is always greater than `U` for any `U` not involving `[element]`, so when forming requirements from rules we might need to swap the two sides, depending on which one is longer once we drop the `[element]`, essentially.

simanerush commented:
@swift-ci Please smoke test macOS

simanerush commented:
@swift-ci please smoke test

simanerush commented:
@swift-ci please smoke test macOS

AnthonyLatsis commented:
@swift-ci please clean smoke test macOS

simanerush commented:
@swift-ci please clean smoke test macOS

simanerush commented:
@swift-ci please clean smoke test

simanerush commented:
@swift-ci please clean smoke test

simanerush committed:
[Requirement Machine] Initialize storage for `ThePackElementSymbol`

simanerush committed:
[Requirement Machine] Add ability to prepend a symbol to a term, clean up existing code

simanerush commented:
@swift-ci please clean smoke test

simanerush commented:
@swift-ci please clean smoke test

simanerush commented:
@martialln Yes, this construct will be allowed🙂

simanerush committed:
[Requirement Machine] When forming requirements from same-element rewrite rules, swap the sides depending on which one is longer once the `[element]` symbol is dropped.

This change fixes the "out-of-order" type parameters error in `GenericSignature::verify`

simanerush commented:
@swift-ci please smoke test macOS

simanerush commented:
@slavapestov Thanks for the feedback! I have refactored the RequirementBuilder changes, and left you a [question](https://github.com/apple/swift/commit/cc9030a9deced5638273d4e2dd74b208c40daba4#r138514042) about concrete same-type requirements. I am also wondering if we still need the [query](https://github.com/apple/swift/pull/70227#discussion_r1468227686) change, given that reordering when forming the requirements worked just fine.

simanerush committed:
[Requirement Machine] Fix the order of terms with pack element symbols to always appear on the left-hand side of the same-element rewrite rule.

simanerush commented:
@swift-ci Please smoke test

AnthonyLatsis commented:
@swift-ci Please smoke test macOS

shaps80 commented:
Amazing to see this making progress. Just curious when we might see this available? Thanks!

simanerush commented:
@shaps80 Currently, this implementation is missing concrete same-element requirements, and unfortunately I can't provide an estimate of when this will be ready.
cc @slavapestov

shaps80 commented:
Ok fair enough. Thanks for letting me know

Lancelotbronner commented:
Will this be part of Swift 6?

vanvoorden commented:
```swift
struct Variadic<each T> {}

extension Variadic where T == {Int, String} {} // {Int, String} is a concrete pack

extension Variadic where T == Never {} // client passed no parameter pack
```

Hi! I'm looking for (someday in the future) using something like this to enable an "optional" parameter pack. Do we think this diff would also implement the "Explicit type pack syntax" from SE-0393[^1] or would that land in another diff? Thanks!

[^1]: https://github.com/swiftlang/swift-evolution/blob/main/proposals/0393-parameter-packs.md#explicit-type-pack-syntax

simanerush commented:
@swift-ci please smoke test

simanerush commented:
@swift-ci please smoke test

simanerush commented:
@swift-ci please smoke test

simanerush commented:
@swift-ci please smoke test macOS

simanerush commented:
@swift-ci please smoke test

dehesa commented:
We are hitting this limitation constantly at [raycast](https://www.raycast.com). We would jump of joy if it could reach Swift 6 by this September 😆

In any case, thank you for working on this.

hborla commented:
>We are hitting this limitation constantly at [raycast](https://www.raycast.com/). We would jump of joy if it could reach Swift 6 by this September

This change will not be included in Swift 6. I understand that people are hitting this limitation frequently and would like for it to be lifted, but there are still open technical challenges with this change. As noted in the PR description, the implementation is gated behind an experimental flag because there are issues with how same-element requirements are represented that we're not happy with. These issues impact the user-facing semantics; they're not just implementation challenges.

hborla commented:
@swift-ci please smoke test

dehesa commented:
I see! Thank you for the openness and setting expectations. I am sad to hear it won’t be there for Swift 6, but if it is not ready, it is not ready 😅
I just wanted to express my strong interest in this feature and my love for the new variadic generics.

slavapestov commented:
@simanerush IMO we should merge this with the user-visible requirement desugaring change behind an experimental flag.
